### PR TITLE
[PERF] web: speed up pager total counter for large DB

### PR DESCRIPTION
**Pull request by fpodoo**
Doing a `search_count` on millions of records could be very slow. On
large databases, the pager counter "1-80 / 8000000" could take several
seconds just to get the total number of records.
    
This commit limits the counter to 10k records, and show this if it
happens: "1-80 / 10000+". If you click on 10000+, it updates with the
real count (or if you go up to 10000 with pager or input value).
 
The search_count on res.partner of odoo.com takes 4s. With this patch,
it will take ~20ms.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/core/pager/pager.js [hunk 1]
```diff
@@ -53,20 +54,30 @@ export class Pager extends Component {
...
      */
     get isSinglePage() {
-        return this.minimum === 1 && this.maximum === this.props.total;
+        return this.minimum === 1 && this.maximum === this.state.total;
+    }
+    /**
+     * Recompute total number of records, if we were at the limit of 10000
+     */
+    async updateCounter() {
+        if (this.state.total == 10000) {
```
Comment (ged-odoo) [inline addons/web/static/src/core/pager/pager.js:63]: use `===` instead of `==`

File: addons/web/static/src/core/pager/pager.js [hunk 2]
```diff
@@ -53,20 +54,30 @@ export class Pager extends Component {
...
     }
-
     /**
      * @param {-1 | 1} direction
      */
     navigate(direction) {
         let minimum = this.props.offset + this.props.limit * direction;
-        if (minimum >= this.props.total) {
+        if ((minimum+this.props.limit) >= this.state.total) {
+            this.updateCounter();
```
Comment (ged-odoo) [inline addons/web/static/src/core/pager/pager.js:73]: probably need to add `await` here, so the rest of the code uses the updated value (so, the `navigate` function need to be `async`

File: addons/web/static/src/core/pager/pager.js [hunk 3]
```diff
@@ -76,7 +87,8 @@ export class Pager extends Component {
      */
     parse(value) {
         let [minimum, maximum] = value.trim().split(/\s*[-\s,;]\s*/);
-        const clamp = (value) => Math.min(Math.max(value, 1), this.props.total);
+        this.updateCounter();
```
Comment (ged-odoo) [inline addons/web/static/src/core/pager/pager.js:90]: as above, this should probably we marked with `await`, so the rest of the code takes the complete value

File: addons/web/static/src/core/pager/pager.xml [hunk 4]
```diff
@@ -11,7 +11,12 @@
                     <span class="o_pager_value d-inline-block border-bottom border-white" t-esc="value" t-on-click="onValueClick" />
                 </t>
                 <span> / </span>
-                <span class="o_pager_limit" t-esc="props.total" />
+                <t t-if="state.total == 10000">
```
Comment (ged-odoo) [inline addons/web/static/src/core/pager/pager.xml:14]: it's fine, but i would personally write it like this:
```xml
    <span class="o_pager_limit" t-on-click.stop="updateCounter">
     <t t-esc="state.total"/><t t-if="state.total===10000">+</t>
  </span>
```

**Review by ged-odoo — COMMENTED** [review]
need some tests :)

File: addons/web/static/src/core/pager/pager.xml (hunk 4 above)
Comment (fpodoo) [reply addons/web/static/src/core/pager/pager.xml:14]: No, in one case there is a t-on-click and not in the other.

File: addons/web/static/src/core/pager/pager.js (hunk 2 above)
Comment (fpodoo) [reply addons/web/static/src/core/pager/pager.js:73]: Not sure about that; the query could be slow (it happens when there are more than 10k records).

Even though it does a rerender, I kept it that way to not slow down the rest of the UI. The pager widget is very small, so the feeling in the UI was quite good. (and it's quite rare; it happens only when you browse after 10k records, and only the first time you pass the limit)

File: addons/web/static/src/core/pager/pager.js (hunk 3 above)
Comment (fpodoo) [reply addons/web/static/src/core/pager/pager.js:90]: Same comment than above.

File: addons/web/models/models.py [hunk 5]
```diff
@@ -66,7 +66,7 @@ def web_search_read(self, domain=None, fields=None, offset=0, limit=None, order=
                 'records': []
             }
         if limit and (len(records) == limit or self.env.context.get('force_search_count')):
-            length = self.search_count(domain)
+            length = self.search_count(domain, limit=10000)
```
Comment (Yenthe666) [inline addons/web/models/models.py:69]: Personally I would really be in favour of adding an `ir.config_parameter` from which this gets the value.
This way more technical partners could fine-tune this number by switching the parameter to another value. There's customers who'd want this number to be much lower and some would like it to be higher :) The difference in performance would be very small while making it totally configurable.

Comment (gustavovalverde) [reply addons/web/models/models.py:69]: +1

**Comment by legalsylvain** [comment]
Nice improvment if it improves performance. 

However, this option should be : 
- 1) configurable (as said by @Yenthe666) (option by model with défault value could be perfect)
- 2) have to implement the possibility to be disabled. (Otherwise it's a régression : in some cases, approximation is not desired)

Thanks !

**Comment by bouvyd** [comment]
Done in #95642

File: addons/web/models/models.py (hunk 5 above)
Comment (inspiredbusiness) [reply addons/web/models/models.py:69]: Or even just a const that another JS module the partner could make could update if needed

**Pull request closed without merging**